KAFKA-17871: avoid blocking the herder thread when producer flushing hangs

davide-armand opened the pull request:
The call to `backingStore.get()` (called by connector task threads
through `OffsetStorageReaderImpl.offsets()`) can block for long time
waiting for data flush to complete (`KafkaProducer.flush()`).

This change moves that call outside the synchronized clause that holds
`offsetReadFutures`, so that if `backingStore.get()` hangs then it does
not keep `offsetReadFutures` locked. The access to `closed` flag
(`closed.get()`) is kept inside the synchronize clause to avoid race
condition with `close()`.

This is important because `OffsetStorageReaderImpl.close()` needs to
lock `offsetReadFutures` as well in order to cancel the futures.  Since
the herder thread calls `OffsetStorageReaderImpl.close()` when
attempting to stops a task, before this change this was resulting in the
herder thread hanging indefinetely waiting for `backingStore.get()` to
complete.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Reviewers: Viktor Somogyi-Vass <viktorsomogyi@gmail.com>, majialong <majialoong@gmail.com>

davide-armand committed:
KAFKA-17871: avoid blocking the herder thread when producer flushing hangs

The call to `backingStore.get()` (called by connector task threads through
`OffsetStorageReaderImpl.offsets()`) can block for long time waiting for data flush to complete
(`KafkaProducer.flush()`).

This change moves that call outside the synchronized clause that holds `offsetReadFutures`,
so that if `backingStore.get()` hangs then it does not keep `offsetReadFutures` locked.
The access to `closed` flag (`closed.get()`) is kept inside the synchronize clause to avoid race
condition with `close()`.

This is important because `OffsetStorageReaderImpl.close()` needs to lock `offsetReadFutures` as
well in order to cancel the futures.
Since the herder thread calls `OffsetStorageReaderImpl.close()` when attempting to stops a task,
before this change this was resulting in the herder thread hanging indefinetely waiting for
`backingStore.get()` to complete.

davide-armand committed:
KAFKA-17871: avoid blocking the herder thread when producer flushing hangs

The call to `backingStore.get()` (called by connector task threads through
`OffsetStorageReaderImpl.offsets()`) can block for long time waiting for data flush to complete
(`KafkaProducer.flush()`).

This change moves that call outside the synchronized clause that holds `offsetReadFutures`,
so that if `backingStore.get()` hangs then it does not keep `offsetReadFutures` locked.
The access to `closed` flag (`closed.get()`) is kept inside the synchronize clause to avoid race
condition with `close()`.

This is important because `OffsetStorageReaderImpl.close()` needs to lock `offsetReadFutures` as
well in order to cancel the futures.
Since the herder thread calls `OffsetStorageReaderImpl.close()` when attempting to stops a task,
before this change this was resulting in the herder thread hanging indefinetely waiting for
`backingStore.get()` to complete.

jlprat commented:
Hi @davide-armand
CI is complaining about check style violations. Could you take a look at it?
Thanks

davide-armand commented:
The CI is now green.
The additional commit that fixed the CI should be squashed before merging.

gharris1727 reviewed (commented):
Hi @davide-armand Thanks for the fix, and thanks for the test!

Are you able to remove the Thread.sleep calls from the test? These could introduce some flakiness later if run on slow machines. Additionally I think this test leaks threads through the executors. Could you ensure that the executors are closed, and the threads terminate?

davide-armand committed:
KAFKA-17871: fixup tests

davide-armand commented:
> Hi @davide-armand Thanks for the fix, and thanks for the test!
> 
> Are you able to remove the Thread.sleep calls from the test? These could introduce some flakiness later if run on slow machines. Additionally I think this test leaks threads through the executors. Could you ensure that the executors are closed, and the threads terminate?

@gharris1727 Pushed a fix in a separate commit, will squash before merging.

I cannot get a green build, if I'm not mistaken it fails on unrelated tests.
Should I keep restarting the CI hoping for a green build?

davide-armand commented:
Rebased on latest `apache:trunk`.
Compilation is now failing, seems unrelated to the changes in the PR.

davide-armand commented:
This PR is still relevant.
I rebased it on latest trunk and got a green build.

viktorsomogyi commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReaderImpl.java:
```diff
@@ -86,14 +86,18 @@ public <T> Map<Map<String, T>, Map<String, Object>> offsets(Collection<Map<Strin
             Future<Map<ByteBuffer, ByteBuffer>> offsetReadFuture;
+
+            // Note: this call can block for long time waiting for data flush to complete (`KafkaProducer.flush()`).
+            offsetReadFuture = backingStore.get(serializedToOriginal.keySet());
```
I'm wondering if we need a behavior change here, afterall the point would be to fix the race condition, but with this change we also call `backingStore.get()` regardless of the value of `closed` which may impose some extra work. Do you think it would make sense to duplicate the if condition below to fail fast?

```suggestion
            if (closed.get()) {
                throw new ConnectException(
                        "Offset reader is closed. This is likely because the task has already been "
                            + "scheduled to stop but has taken longer than the graceful shutdown "
                            + "period to do so.");
            }
            
            // Note: this call can block for long time waiting for data flush to complete (`KafkaProducer.flush()`).
            offsetReadFuture = backingStore.get(serializedToOriginal.keySet());
```

viktorsomogyi replied:
(Ofc with the error message moved to a constant to avoid duplication)

davide-armand replied:
@viktorsomogyi  I forced pushed the suggested change, I agree it makes sense to fail fast here.
I also pushed a separate commit to refactor `OffsetStorageReaderImpl.offsets()` into smaller methods to fix a checkstyle complexity error in the build caused by this change.

davide-armand committed:
KAFKA-17871: Refactoring to fix checkstyle NPath complexity error

Refactor `OffsetStorageReaderImpl.offsets()` by splitting it into
methods for serialization, backing store read, and deserialization.
This is for lowering the NPath complexity of the offsets() method which
was causing the build to fail.

viktorsomogyi commented:
@davide-armand thank you for your contribution!

majialoong commented:
Hi @davide-armand @viktorsomogyi , would you mind reviewing this PR https://github.com/apache/kafka/pull/22632 when you have a chance? This test has been showing up as flaky fairly often recently, so it would be great to get your input. Thanks!
